pkg/controllers/hub/membercluster/controller.go

```diff
@@ -0,0 +1,129 @@
+			klog.V(4).InfoS("Ignoring NotFound memberCluster", "memberCluster", mcObjRef)
+			return ctrl.Result{}, nil
+		}
+		klog.ErrorS(err, "Failed to get memberCluster", "memberCluster", mcObjRef)
+		return ctrl.Result{}, err
+	}
+	// Handle deleting member cluster, removes finalizers on all the resources in the cluster namespace
+	// after member cluster force delete wait time.
+	if !mc.DeletionTimestamp.IsZero() && time.Since(mc.DeletionTimestamp.Time) >= r.ForceDeleteWaitTime {
+		klog.V(2).InfoS("The member cluster deletion is stuck removing the "+
+			"finalizers from  all the resources in member cluster namespace", "memberCluster", mcObjRef)
+		return r.removeFinalizer(ctx, mc)
+	}
+
+	return ctrl.Result{RequeueAfter: r.ForceDeleteWaitTime}, nil
+}
+
+// removeFinalizer removes finalizers on the resources in the member cluster namespace.
+// For EndpointSliceExport, InternalServiceImport & InternalServiceExport resources, the finalizers should be
+// removed by other hub networking controllers when leaving. So this MemberCluster controller only handles
+// EndpointSliceImports here.
+func (r *Reconciler) removeFinalizer(ctx context.Context, mc clusterv1beta1.MemberCluster) (ctrl.Result, error) {
+	// Remove finalizer for EndpointSliceImport resources in the cluster namespace.
+	mcObjRef := klog.KRef(mc.Namespace, mc.Name)
+	mcNamespace := fmt.Sprintf(hubconfig.HubNamespaceNameFormat, mc.Name)
+	var endpointSliceImportList fleetnetv1alpha1.EndpointSliceImportList
+	if err := r.Client.List(ctx, &endpointSliceImportList, client.InNamespace(mcNamespace)); err != nil {
+		klog.ErrorS(err, "Failed to list endpointSliceImports", "memberCluster", mcObjRef)
+		return ctrl.Result{}, err
+	}
+	errs, ctx := errgroup.WithContext(ctx)
+	for i := range endpointSliceImportList.Items {
+		esi := &endpointSliceImportList.Items[i]
+		errs.Go(func() error {
+			esiObjRef := klog.KRef(esi.Namespace, esi.Name)
+			esi.SetFinalizers(nil)
+			if err := r.Client.Update(ctx, esi); err != nil {
+				klog.ErrorS(err, "Failed to remove finalizers for endpointSliceImport",
+					"memberCluster", mcObjRef, "endpointSliceImport", esiObjRef)
+				return err
+			}
+			klog.V(2).InfoS("Removed finalizers for endpointSliceImport",
+				"memberCluster", mcObjRef, "endpointSliceImport", esiObjRef)
+			return nil
+		})
+	}
+	return ctrl.Result{}, errs.Wait()
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	customPredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			// Ignore creation events.
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// trigger reconcile on delete event just in case update event is missed.
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// If new object is being deleted, trigger reconcile.
+			if e.ObjectOld.GetDeletionTimestamp() == nil && e.ObjectNew.GetDeletionTimestamp() != nil {
```

Comment:
can you validate that if we hit a error or requeue the request, will the request be ignored because of this UpdateFunc?

Reply:
after thorough consideration, i think it's better not to check the e.ObjectOld.GetDeletionTimestamp() == nil or not, it's a middle state. The philosophy is to reconcile based on the desired state.